Add persistence

rrostt opened the pull request:
Add persist option. Default is false.

When true data fetched from a given url will persist in localStorage for 24h.

Note: when data is not fetched using fetch it is not sent through interceptors.

alex-cory commented:
Will review this asap.

alex-cory commented:
Suggestion 1 might look something like
```js

```

rrostt commented on package.json:
```diff
@@ -9,6 +9,7 @@
     "url": "https://github.com/alex-cory/use-http.git"
   },
   "dependencies": {
+    "eslint-plugin-jest-formatting": "^1.2.0",
```
doh, of course. problem was it is it not in devDependencies either which is why I added it.

alex-cory replied:
just move to `devDependencies`. I just realized it was missing for some weird reason.

rrostt commented:
Excellent. Will update asap. Hadn't really looked at doFetchArgs but it makes perfect sense to move it there!

Also, not sure if you're aware of the two lint errors?

alex-cory commented:
@mogelbrod yes, I have thought about this. I haven't had time to finish `use-react-storage`, but this will solve this issue using the same syntax for both `localStorage` and `AsyncStorage`

rrostt commented:
Did an overhaul. Have a look. Have not added isBrowser checks yet. Tests pass.

mogelbrod commented on src/doFetchArgs.ts:
```diff
@@ -105,7 +107,10 @@ export default async function doFetchArgs<TData = any>(
       cached: cache.get(responseID) as Response | undefined,
-      ageID: responseAgeID
+      ageID: responseAgeID,
+      age: (cache.get(responseAgeID) || 0) as number,
```
`age` was removed from response in favour of `isExpired`.

mogelbrod commented on src/persistentStorage.ts:
```diff
@@ -0,0 +1,28 @@
+const cacheName = 'useHTTPcache'
+
+const hasPersistentData = (url: string): boolean => {
+  const urlCache = JSON.parse(localStorage[cacheName] || '{}')
```
I think error handling is needed for when parsing fails (probably fallback to empty cache)

rrostt commented:
Made some updates. See commits.

alex-cory commented:
@rrostt This is great so far! I made a [branch](https://github.com/rrostt/use-http/compare/add-persistence...alex-cory:add-persistence) to show you what I was thinking. This is super close though and thanks for all your hard work!

rrostt commented:
@alex-cory yeah, i was considering just making the persistancestorage one type of cache and then do exactly like you show here with inmemory map or localstorage cache. the reason i did not do this was because to do that we must replicate Response and i did not want to introduce that much complexity at this stage.

basically we must create a serialisable version of Response and store that. i'm not sure that is trivial? looking at it quickly though, it looks like the current implementation only needs to support the `ok` property and `json` and `text` methods essentially. I can easily create an object from a response object that implements those and that is serialisable that we can then store in our cache (both memory and localstorage, etc).

any comments?

alex-cory commented:
@rrostt I agree. Putting the Response in localStorage is not trivial, but it's important. I think the most important things are `ok`, `status`, `statusText`, `headers` and `body`.
```js
var a = new Response(JSON.stringify({ no: 'way' }), {
  status: 200,
  statusText: '',
  headers: new Headers({ 'Content-Type': 'application/json' })
})
var test = async () => {
    console.log(await a.json())
}
test()
```
This seems to work^

If you find a different way, totally open to it.

rrostt commented:
Cool, will take a stab tonight unless I get corona. Cheers!

alex-cory commented:
LOL

rrostt commented:
Did not get corona. Have a look, I pulled from your branch as I liked what you'd done, and then continued from there.
I removed dealing with expiration of cache to the individual caches. Not sure if this will mess with future cache policies? The result is a simplified useFetch, and doFetchArgs.

alex-cory commented on src/useCache.ts:
```diff
@@ -0,0 +1,108 @@
+const useCache = ({ persist, cacheLife }: UseCacheArgs): Cache => {
+  const { isNative, isServer, isBrowser } = useSSR()
+  invariant(!(isServer && persist), 'There is no persistant storage on the Server currently! 🙅‍♂️')
+  invariant(!(isNative && !isServer && !isBrowser), 'React Native support is not yet implemented!')
```
We should update this to be `'React Native support for persistant cache is not yet implemented!'`

alex-cory commented on src/useCache.ts:
```diff
@@ -0,0 +1,108 @@
+}
+
+const inMemoryStorage = new Map<string, Response | number>()
+const getMemoryStorage = ({ cacheLife }: { cacheLife: number }): Cache => ({
```
This is AWESOME! So much cleaner 🤩

rrostt commented:
Did some testing with the way we used toResponseObject to serialise Response object, and it was a no go. It worries me that the tests went fine. Somewhere some implementation is not accurate, testing-library?

So I changed it and created a simpler serialiseResponse method that extracts body, status, statusText, and headers. This is what we use to recreate the Response object with.

I also refactored useCache and moved localstorage and memorystorage to separate modules to make them testable. Added tests for localstorage (and found some bugs doing so which I corrected).

It would seem as if things are coming together nicely...

alex-cory commented on src/__tests__/useFetch.test.tsx:
```diff
@@ -803,3 +805,68 @@ describe('useFetch - BROWSER - errors', (): void => {
+
+  it('should fetch once', async (): Promise<void> => {
+    const { waitForNextUpdate } = renderHook(
+      () => useFetch({ url: 'https://persist.com', persist: true, cachePolicy: NO_CACHE }, [])
```
we need to error if we're trying to persist data with `NO_CACHE` cachePolicy.

alex-cory commented on README.md:
```diff
@@ -700,6 +701,9 @@ const options = {
   // The time in milliseconds that cache data remains fresh.
   cacheLife: 0,
 
+  // Sets wether to persist data ot not after page refresh. Only available on Browser.
```
Sets whether to persist the cache after page refresh. Only available in Browser.

alex-cory replied:
`// allows caching to persist after page refresh. Only available in Browser.`

alex-cory commented on src/useFetch.ts:
```diff
@@ -70,31 +71,25 @@ function useFetch<TData = any>(...args: UseFetchArgs): UseFetch<TData> {
-            error.current = err
-            setLoading(false)
-          }
+      if (response.isCached && (persist || cachePolicy === CACHE_FIRST)) {
```
I don't think we need the `persist` here. This would mean NO_CACHE or another cachePolicy without a cache would pass here if `persist` is true.

alex-cory commented:
Also, need to add test to make sure the `response` has the right values.  While I was playing around with this, there was no `response.ok`, there was only `response.data`. You can see my fix for this in my last commit above.

rrostt commented:
Excellent, good job!

Do you want me to do anything more on this?

alex-cory commented:
I'll clean up the rest today and if I need more help, will let you know!

alex-cory commented:
**Things left to complete:**
1. We need to figure out which `cachePolicy`s cannot have `persist: true` with them. Then create `invariants` for them. [See comment](https://github.com/alex-cory/use-http/pull/199#discussion_r392712347)
2. We need to figure out which `cachePolicy`s can return from cache in the same way as `cache-first` does. Then add them to [this line](https://github.com/rrostt/use-http/blob/add-persistence/src/useFetch.ts#L84) and remove the `persist` from that if block.
3. need to add test to make sure the `response` has the right values. [See comment](https://github.com/alex-cory/use-http/pull/199#issuecomment-599286707)

rrostt commented:
for 3) how did you test this? Currently our local storage cache only returns a Response-object. It should not be possible for it not to have the ok-property, unless there is an error in the underlying implementation of Response. (e.g. node-fetch implements ok as a get property checking status >= 200 and < 300.)

alex-cory commented:
@rrostt I added [this line](https://github.com/alex-cory/use-http/pull/199/commits/b12ba33a6104d9bdbe6feabb4101208f2f374736#diff-326372ad77dca3e7c2460c8d26329f41R86). In order for the `response` to have all the fields in the return from `const { response } = useFetch()` we have to set `res.current = response.cached` so it actually has the `Response` in it. Otherwise we were only setting the `data` field on it.

alex-cory commented on src/useFetch.ts:
```diff
@@ -114,9 +110,8 @@ function useFetch<TData = any>(...args: UseFetchArgs): UseFetch<TData> {
-        if (cachePolicy === CACHE_FIRST) {
-          cache.set(response.id, newRes.clone())
-          if (cacheLife > 0) cache.set(response.ageID, Date.now())
+        if (persist || cachePolicy === CACHE_FIRST) {
```
Hmm... I think having just the `cachePolicy` check here should work. If `persist` is set with the incorrect `cachePolicy` an Invariant Violation will already be thrown.

rrostt replied:
I agree. I originally had a different idea about the persistence, but you're right.
